core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -424,79 +466,47 @@ protected boolean checkBlocks() {
                 }
                 return false;
             }
+            // request does not have a shardId yet, we need to pass the concrete index to resolve shardId
+            resolveRequest(clusterState.metaData(), concreteIndex, request);
+            assert request.shardId() != null : "request shardId must be set in resolveRequest";
             return true;
         }
 
-        protected ShardRouting resolvePrimary(ShardIterator shardIt) {
-            // no shardIt, might be in the case between index gateway recovery and shardIt initialization
-            ShardRouting shard;
-            while ((shard = shardIt.nextOrNull()) != null) {
-                // we only deal with primary shardIt here...
-                if (shard.primary()) {
-                    return shard;
-                }
-            }
-            return null;
-        }
+        private void performAction(final String nodeId, final String action, final boolean isPrimaryAction) {
+            DiscoveryNode node = observer.observedState().nodes().get(nodeId);
+            transportService.sendRequest(node, action, request, transportOptions, new BaseTransportResponseHandler<Response>() {
```

Comment:
side note: the whole point of the observer was to sample the cluster at known moments and be consistent. I agree the new code is simpler so change is good. @jasontedor if you have a better suggestion for doc/naming in the cluster state observer please let me know.